shell/platform/fuchsia/flutter/software_surface_producer.cc

```diff
@@ -0,0 +1,236 @@
+                   });
+  if (exact_match_it != available_surfaces_.end()) {
+    TRACE_EVENT_INSTANT0("flutter", "Exact match found");
+    surface = std::move(*exact_match_it);
+    available_surfaces_.erase(exact_match_it);
+  } else {
+    surface = CreateSurface(size);
+  }
+
+  if (surface == nullptr) {
+    FML_LOG(ERROR) << "SoftwareSurfaceProducer: Could not acquire surface";
+    return nullptr;
+  }
+
+  if (!surface->FlushSessionAcquireAndReleaseEvents()) {
+    FML_LOG(ERROR)
+        << "SoftwareSurfaceProducer: Could not flush acquire/release "
+           "events for buffer.";
+    return nullptr;
+  }
+
+  return surface;
+}
+
+void SoftwareSurfaceProducer::SubmitSurfaces(
+    std::vector<std::unique_ptr<SurfaceProducerSurface>> surfaces) {
+  TRACE_EVENT0("flutter", "SoftwareSurfaceProducer::SubmitSurfaces");
+
+  // Submit surface
+  for (auto& surface : surfaces) {
+    SubmitSurface(std::move(surface));
+  }
+
+  // Buffer management.
+  AgeAndCollectOldBuffers();
+}
+
+void SoftwareSurfaceProducer::SubmitSurface(
+    std::unique_ptr<SurfaceProducerSurface> surface) {
+  TRACE_EVENT0("flutter", "SoftwareSurfacePool::SubmitSurface");
+  FML_CHECK(valid_);
+
+  // This cast is safe because |SoftwareSurface| is the only implementation of
+  // |SurfaceProducerSurface| for Flutter on Fuchsia.  Additionally, it is
+  // required, because we need to access |SoftwareSurface| specific information
+  // of the surface (such as the amount of memory it contains).
+  auto software_surface = std::unique_ptr<SoftwareSurface>(
+      static_cast<SoftwareSurface*>(surface.release()));
+  if (!software_surface) {
+    return;
+  }
+
+  uintptr_t surface_key = reinterpret_cast<uintptr_t>(software_surface.get());
+  auto insert_iterator = pending_surfaces_.insert(std::make_pair(
+      surface_key,                 // key
+      std::move(software_surface)  // value
+      ));
+  if (insert_iterator.second) {
+    insert_iterator.first->second->SignalWritesFinished(std::bind(
+        &SoftwareSurfaceProducer::RecyclePendingSurface, this, surface_key));
+  }
+}
+
+std::unique_ptr<SoftwareSurface> SoftwareSurfaceProducer::CreateSurface(
+    const SkISize& size) {
+  TRACE_EVENT2("flutter", "SoftwareSurfacePool::CreateSurface", "width",
+               size.width(), "height", size.height());
+  auto surface =
+      std::make_unique<SoftwareSurface>(sysmem_allocator_, flatland_allocator_,
+                                        scenic_session_, size, buffer_id_++);
+  if (!surface->IsValid()) {
+    FML_LOG(ERROR) << "SoftwareSurfaceProducer: Created surface is invalid";
+    return nullptr;
+  }
+  trace_surfaces_created_++;
+  return surface;
+}
+
+void SoftwareSurfaceProducer::RecycleSurface(
+    std::unique_ptr<SoftwareSurface> surface) {
+  // The surface may have become invalid (for example it the fences could
+  // not be reset).
+  if (!surface->IsValid()) {
+    return;
```

Comment:
Even if the root cause was elsewhere, should we log an error here anyways?

Reply:
Done